platforms/jvm/plugins-java-base/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -304,47 +305,23 @@ private void defineConfigurationsForSourceSet(SourceSet sourceSet, RoleBasedConf
 
     private void configureCompileDefaults(final Project project, final DefaultJavaPluginExtension javaExtension) {
         project.getTasks().withType(AbstractCompile.class).configureEach(compile -> {
-            ConventionMapping conventionMapping = compile.getConventionMapping();
-            conventionMapping.map("sourceCompatibility", () -> computeSourceCompatibilityConvention(javaExtension, compile).toString());
-            conventionMapping.map("targetCompatibility", () -> computeTargetCompatibilityConvention(javaExtension, compile).toString());
+            JvmPluginsHelper.configureCompileDefaults(compile, javaExtension, (@Nullable JavaVersion rawConvention, Supplier<JavaVersion> javaVersionSupplier) -> {
+                if (compile instanceof JavaCompile) {
```

Comment:
I don't think we can make this change. Users can (and do) override `AbstractCompile` -- whether they should or should not is a different question. This would be a behavior change

The java base plugin is in reality the JVM base plugin

Reply:
I'm going to argue to keep this change.  The `java/jvm-base` plugin configures the defaults for instances of `JavaCompile` and sets convention defaults on non-`JavaCompile`-typed `AbstractCompile` tasks.  This is all it should know about.

To make the `GroovyCompile` type available, you have to apply the `groovy` plugin, which applies `groovy-base`.  `groovy-base`'s application will catch every instance of `GroovyCompile` tasks and configure them properly.

Extending `AbstractCompile` is fine - you'll get picked up by this loop.  `GroovyCompile` also extends `AbstractCompile`, so it would have hit the original `if` condition, not the `else` anyway.  

The rest is just reorganization and de-duplication.  This change keeps `JavaBasePlugin` from needing to know about `GroovyCompile`, which is a good separation we should have.

Can you explain why you think this could be a behavior change?  I don't see it - unless you do something to get access to the `GroovyCompile` type without applying the `groovy` plugins, in which case you're now your own plumber.

Reply:
Well. Yeah that's pretty much it. Previously GroovyCompile would get its defaults without applying the groovy base plugin. Now you need to apply the groovy base plugin to get those defaults.

Not saying that's not where we should be moving, but that is a breaking behavior change right now

Reply:
Perhaps I have missed something here. It looks like we are still setting source/target compatibility for all `AbstractCompile` instances. 

But, whats with removing the `ompile.getDestinationDirectory().convention(project.getProviders().provider(new BackwardCompatibilityOutputDirectoryConvention(compile)));`? I don't see that anywhere else in this PR

Reply:
Please excuse me. I thought you updated this logic to only apply the to `JavaCompile` instances. I didn't realize this logic was in a _callback_ and already was performing this check elsewhere. 

Same question applies about `BackwardCompatibilityOutputDirectoryConvention` though

Reply:
Okay. Read this over _again_. I think the use-case above that I wrote still applies. If the user was using `java-base` and `GroovyCompile`, they will no longer get the defaults. They would now need to apply `groovy` to get those defaults

Reply:
We spoke in depth about this and landed on the current approach.